src/main/java/com/google/googleidentity/oauth2/endpoint/TokenEndpoint.java

```diff
@@ -0,0 +1,139 @@
+ */
+@Singleton
+public class TokenEndpoint extends HttpServlet {
+
+    private static final long serialVersionUID = 5L;
+
+    private static final Logger log = Logger.getLogger("TokenEndpoint");
+
+    private final ClientDetailsService clientDetailsService;
+
+
+    @Inject
+    public TokenEndpoint(ClientDetailsService clientDetailsService) {
+        this.clientDetailsService = clientDetailsService;
+    }
+
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException, UnsupportedOperationException {
+
+        OAuth2Exception exception =
+                new InvalidRequestException(
+                        InvalidRequestException.ErrorCode.UNSUPPORTED_REQUEST_METHOD);
+        log.info("Token endpoint does not support GET request." );
+        response.setStatus(exception.getHttpCode());
+        response.setContentType("application/json;charset=UTF-8");
+        response.getWriter().println(
+                OAuth2ExceptionHandler.getResponseBody(exception).toJSONString());
+        response.getWriter().flush();
+    }
+
+    protected void doPost(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException, UnsupportedOperationException {
+        try {
+            TokenEndpointRequestValidator.validatePost(request);
+        } catch(OAuth2Exception exception) {
+            log.info(
+                    "Failed in validating Post request in Token Endpoint." +
+                            "Error Type: " + exception.getErrorType() +
+                            "Description: " + exception.getErrorDescription());
+            response.setStatus(exception.getHttpCode());
+            response.setContentType("application/json;charset=UTF-8");
+            response.getWriter().println(
+                    OAuth2ExceptionHandler.getResponseBody(exception).toJSONString());
+            response.getWriter().flush();
+            return;
+        }
+
+        OAuth2Request oauth2Request = parseOAuth2RequestFromHttpRequest(request);
+    }
+
+    /**
+     * Should be called after all validation in doPost function.
+     */
+    private OAuth2Request parseOAuth2RequestFromHttpRequest(HttpServletRequest request) {
+
+        String grantType = request.getParameter(OAuth2ParameterNames.GRANT_TYPE);
+
+        Preconditions.checkArgument(
+                OAuth2Utils.getClientSession(request).getClient().isPresent(),
+                "Client should have been set in client filter!");
+
+        OAuth2Request.Builder oauth2RequestBuilder = OAuth2Request.newBuilder();
+        oauth2RequestBuilder.getRequestAuthBuilder().setClientId(
+                            OAuth2Utils.getClientSession(request).getClient().get().getClientId());
+        oauth2RequestBuilder.getRequestBodyBuilder()
+                .setGrantType(grantType).setResponseType("token");
+
+        if (grantType.equals("authorization_code")) {
```

Comment:
Grant types are used in a handful of places in the codebase. Can you extract constants to a separate class for grant types like `"authorization_code"` and `"token"`? Followup pull request is fine if you want.